dnf/cli/commands/history.py

```diff
@@ -219,24 +243,140 @@ def str2transaction_id(s):
+            nas = subj.get_nevra_possibilities(forms=[hawkey.FORM_NA])
+
+            if len(nas) != 1:
+                raise dnf.exceptions.Error(_('Cannot parse name.arch for package "{na}".').format(na=na))
+
+            parsed_na = nas[0]
+            try:
+                action = pkg_data["action"]
+
+                # validate the reasons before running the transaction and cache them
+                try:
+                    self._na_reason_cache[na] = \
+                        libdnf.transaction.StringToTransactionItemReason(pkg_data["reason"])
+                except IndexError as e:
+                    raise dnf.exceptions.Error(
+                        _('Unexpected value of package reason "{reason}" for package "{pkg}" in "{fn}".')
+                            .format(reason=pkg_data["reason"], pkg=na, fn=fn)
+                    )
+
+                if action in ("Install", "Upgrade", "Downgrade", "Reinstall"):
+                    nevra = "%s-%s.%s" % (parsed_na.name, pkg_data["to"], parsed_na.arch)
+                    if action in ("Upgrade", "Downgrade", "Reinstall"):
+                        from_nevra = "%s-%s.%s" % (parsed_na.name, pkg_data["from"], parsed_na.arch)
+                        query = dnf.subject.Subject(from_nevra).get_best_solution(self.base.sack)["query"]
+                        if not query.installed():
+                            raise dnf.exceptions.Error(
+                                _('Package "{pkg}" not found installed for action "{action}" in "{fn}".')
+                                    .format(pkg=from_nevra, action=action, fn=fn)
+                            )
+
+                elif action == "Removed":
+                    nevra = "%s-%s.%s" % (parsed_na.name, pkg_data["from"], parsed_na.arch)
+                else:
+                    raise dnf.exceptions.Error(
+                        _('Unexpected value of package action "{action}" for package "{pkg}" in "{fn}".')
+                            .format(action=action, pkg=na, fn=fn)
+                    )
+
+                query = dnf.subject.Subject(nevra).get_best_solution(self.base.sack)["query"]
+
+                if not query:
+                    raise dnf.exceptions.Error(_('Package "{nevra}" from "{fn}" was not found.')
+                        .format(nevra=nevra, fn=fn))
+
+                if action in ("Install", "Upgrade", "Downgrade"):
+                    sltr = dnf.selector.Selector(self.base.sack).set(pkg=query)
+                    self.base.goal.install(select=sltr, optional=False)
+                elif action == "Reinstall":
+                    sltr = dnf.selector.Selector(self.base.sack).set(pkg=query.available())
+                    self.base.goal.install(select=sltr, optional=False)
+                elif action == "Removed":
+                    self.base.goal.erase(query.installed()[0])
+                elif action == "Obsolete":
+                    raise RuntimeError("what about obsolete?")
```

Comment:
I am sorry, is it a joke? We cannot merge it with out support of obsoletes.

Reply:
:sunglasses: obviously it's one of the missing things...